storage/engine/mvcc.go

```diff
@@ -924,19 +924,97 @@ func MVCCDeleteRange(engine Engine, ms *MVCCStats, key, endKey proto.Key, max in
 	return num, nil
 }
 
+func getScanMetaKey(iter Iterator, encEndKey proto.EncodedKey) (proto.Key, proto.EncodedKey, error) {
+	metaKey := iter.Key()
+	if bytes.Compare(metaKey, encEndKey) >= 0 {
+		if err := iter.Error(); err != nil {
+			return nil, nil, err
+		}
+		return nil, nil, nil
+	}
+	key, _, isValue := MVCCDecodeKey(metaKey)
+	if isValue {
+		return nil, nil, util.Errorf("expected an MVCC metadata key: %q", metaKey)
+	}
+	return key, metaKey, nil
+}
+
+func getReverseScanMetaKey(iter Iterator, encEndKey proto.EncodedKey) (proto.Key, proto.EncodedKey, error) {
+	metaKey := iter.Key()
+	// The metaKey < encEndKey is exceeding the boundary.
+	if bytes.Compare(metaKey, encEndKey) < 0 {
+		if err := iter.Error(); err != nil {
+			return nil, nil, err
+		}
+		return nil, nil, nil
+	}
+
+	// The row with oldest version will be got by seeking reversely. We use the
+	// key of this row to get the MVCC metadata key.
+	key, _, isValue := MVCCDecodeKey(metaKey)
+	if isValue {
+		var keyBuf []byte
+		iter.Seek(mvccEncodeKey(keyBuf, key))
+		if !iter.Valid() {
+			if err := iter.Error(); err != nil {
+				return nil, nil, err
+			}
+			return nil, nil, nil
+		}
+
+		metaKey = iter.Key()
+		// The metaKey < encEndKey is exceeding the boundary.
+		if bytes.Compare(metaKey, encEndKey) < 0 {
+			if err := iter.Error(); err != nil {
+				return nil, nil, err
+			}
+			return nil, nil, nil
+		}
+		_, _, isValue = MVCCDecodeKey(metaKey)
+		if isValue {
+			return nil, nil, util.Errorf("expected an MVCC metadata key: %q", metaKey)
+		}
+	} else {
+		//The inline rows don't hava versioned values.
+	}
+	return key, metaKey, nil
+}
+
 // MVCCScan scans the key range specified by start key through end key
 // up to some maximum number of results. Specify max=0 for unbounded
 // scans.
 func MVCCScan(engine Engine, key, endKey proto.Key, max int64, timestamp proto.Timestamp,
 	consistent bool, txn *proto.Transaction) ([]proto.KeyValue, []proto.Intent, error) {
 	res := []proto.KeyValue{}
-	intents, err := MVCCIterate(engine, key, endKey, timestamp, consistent, txn, func(kv proto.KeyValue) (bool, error) {
-		res = append(res, kv)
-		if max != 0 && max == int64(len(res)) {
-			return true, nil
-		}
-		return false, nil
-	})
+	intents, err := MVCCIterate(engine, key, endKey, timestamp, consistent, txn, false,
```

Comment:
try replacing `false` by `false /* reverse */`. That way the reader knows what it does.